### PR TITLE
document (and simplify) release process

### PR DESCRIPTION
**Pull request by igalic**

### CODE REVIEW

**Comment by igalic** [comment]
@carlossg do you think this ^ makes sense / will work?

yeah, i could burn a version number to try it out, but i'm also thinking, maybe this task is better put elsewhere? or maybe it exists already somewhere?

File: README.markdown [hunk 1]
```diff
@@ -92,6 +92,20 @@ was introduced.
 
 ##Development
 
+We highly welcome contributions to this module. We like new contributions to
+include include documentation, and rspec tests, but will happily guide you
```
Comment (raphink) [inline README.markdown]: include include

**Comment by raphink** [comment]
That's nice. Just one thing I don't really like with using Blacksmith locally: there's no CHANGELOG.md currently, and I think it would be good to have one, but then you'd need to update it before doing the `rake lint_release` call…

**Comment by liamjbennett** [comment]
This looks good to me.

@igalic I tired adding the deploy section to the .travis file on windows_autoupdate and I am getting "Username/password do not match" - anything I am missing?

**Comment by igalic** [comment]
@nibalizer ping ^

**Comment by raphink** [comment]
@liamjbennett did you reencrypt the password for your repo?

**Comment by igalic** [comment]
@raphink i'm not sure, re CHANGELOG.md :s
i think the main problem with that is, that changelogs are *actually* hard. hard to maintain. hard to maintain with sensible, human relevant information…

**Comment by raphink** [comment]
Yes, but they are displayed on the forge and actually useful.

**Comment by igalic** [comment]
*nod* well, in that case i'm glad i involved @carlossg in this discussion, maybe he's got something smart to say about how to do this :P

**Comment by raphink** [comment]
I think that it's hard to generate a relevant CHANGELOG file dynamically. What is easier however, is to prepare it every time a relevant change is made.

**Comment by raphink** [comment]
This is what we do in Augeas https://github.com/hercules-team/augeas/blob/master/NEWS where NEWS is updated as it goes (ideally).

Another option, which we try to do at @camptocamp with our Puppet modules, is to release every time there is a significant change, and thus make changes to CHANGELOG every time.

**Comment by ffrank** [comment]
+1 for *having* change logs. Bothering users to parse git logs or even compare READMEs is a nono.

+1 also for being scared of the burden of maintaining them.

Perhaps we should have requirements not unlike puppetlabs/puppet e.g.
- [ ] branch features at least one commit that starts with the issue number
- [ ] said commit message uses imperative
- [ ] branch includes an update to the changelog in the *unreleased* section on top
etc.

**Comment by igalic** [comment]
i aggree with many of these. [npm has of course solved](https://www.npmjs.com/package/semantic-release) this issue already:

KILL ALL HUMANS

and the way they solved is kinda what @ffrank suggests, having remotely parsable git-logs, which can generate both, a SemVer compatible version, and a close to readable changelog…

**Comment by raphink** [comment]
Yep, I totally agree. Is there a tool that generates a markdown changelog from such conventions in git commits?

**Comment by ffrank** [comment]
PL *does* generate change logs, I think by collecting the (PUP-xxxx) tags and relating them to Jira.

However, my third bullet point is actually me suggesting that we can just require contributors to keep the log updated by policy.

**Comment by igalic** [comment]
versioning is hard and shouldn't be done by humans.
can we not just have regression tests against our api/abi?
which shout at you when you break it, and we decide on case-by-case if the breakage is wroth it?

oh, wait, that's the point of tests, right? kinda…

**Comment by carlossg** [comment]
Maybe you want a module:release with an option `build => false` so it does everything but the build and push to the forge?

I would encourage git logs that are useful for humans and parseable into the changelog. The release task could ensure the changelog is updated.

**Comment by nibalizer** [comment]
Changelogs++ also I dont like much git commit as a changelog, especially around reverts. IMHO changelogs should be written up by a human. Not that I am very good about doing this with my own software.

**Comment by carlossg** [comment]
Git commits are written by humans too (mostly) but only a few times *for* humans :)

**Comment by ffrank** [comment]
I'd be fine with a rule of "If it matches `^#\d+: ` include it in the change log at release" or somesuch. Agree with @carlossg that this *will* likely be easier to assert than constant changes to the `Changelog.md` proper.

**Comment by nibalizer** [comment]
@ffrank I don't think most modules will have bugs attached. Most PRs on modules is just 'hey here is feature'

**Comment by ffrank** [comment]
Well, what Puppet does is say "well, thanks, but please also file a ticket" :-)

It's likely more practical to have an alternate pattern like `/^NEW:/`

**Comment by igalic** [comment]
okay. here's my proposal:

changelogs are super important for humans, robots and imaginary internet points in the forge (thanks for this line, @cmurphy ;) but our release task should only care that there ***is*** a mention of the current release in CHANGELOG.md

**Comment by nibalizer** [comment]
@igalic +1

File: Rakefile [hunk 2]
```diff
@@ -53,3 +53,17 @@ task :test => [
   :lint,
   :spec,
 ]
+
+desc "Offload release process to Travis."
+task :travis_release => [
+  :check_changelog,  # check that the changelog contains an entry for the current release
+  :"module:release"[ :build => false], # do everything except build / push to forge, travis will do that for us
```
Comment (igalic) [inline Rakefile]: @carlossg is this ^ the right way to do it?

Comment (igalic) [inline Rakefile]: i guess that's a no

```
$ bundle exec rake test
rake aborted!
TypeError: can't convert Hash into Integer
/home/travis/build/puppet-community/puppetcommunity-extlib/Rakefile:60:in `[]'
/home/travis/build/puppet-community/puppetcommunity-extlib/Rakefile:60:in `<top (required)>'
/home/travis/.rvm/gems/ruby-1.9.3-p551/bin/ruby_executable_hooks:15:in `eval'
/home/travis/.rvm/gems/ruby-1.9.3-p551/bin/ruby_executable_hooks:15:in `<main>'
(See full trace by running task with --trace)
```

Comment (carlossg) [inline Rakefile]: You can define now

```
    Blacksmith::RakeTask.new do |t|
      t.build = false # do not build the module nor push it to the Forge, just do the tagging [:clean, :tag, :bump_commit]
    end
```

then just 

```
task :travis_release => [
  :check_changelog,  # check that the changelog contains an entry for the current release
  :"module:release"]
```

**Comment by igalic** [comment]
great. ruby 1.8 failing because
```
Gem::InstallError: mime-types requires Ruby version >= 1.9.2.
```

**Comment by carlossg** [comment]
You can force mime-types < 2.0, for 1.8 compatibility

**Comment by igalic** [comment]
done, next up:
```
Gem::InstallError: rest-client requires Ruby version >= 1.9.2.

An error occurred while installing rest-client (1.8.0), and Bundler cannot continue.
```
which i find extra ironic due to its version:
```
Make sure that `gem install rest-client -v '1.8.0'` succeeds before bundling.
```

**Comment by carlossg** [comment]
Actually, blacksmith does not work in 1.8 so just make the gem install in ruby 1.9+
As you won't be running it from 1.8 it shouldn't be a problem

**Comment by igalic** [comment]
@carlossg but how do i exclude the rake-task in 1.8?

**Comment by carlossg** [comment]
just wrap the tasks with `if RUBY_VERSION >= "1.9.0"` in the Rakefile

**Comment by igalic** [comment]
well, this is ugly, but now it works, and i'll merge it cuz i can't see it anymore :P

**Comment by carlossg** [comment]
supporting 1.8, that *is* ugly ;)

**Pull request merged**